ugvc/pipelines/deepvariant/training_set_consistency_check.py

```diff
@@ -0,0 +1,237 @@
+    parser.add_argument(
+        "--normal_crams",
+        type=str,
+        help="crams of normal samples, used for training somatic models",
+    )
+    parser.add_argument(
+        "--min_count_snps", type=int, default=3, help="number of snp supporting reads to count as a ground-truth hit"
+    )
+    parser.add_argument(
+        "--min_count_germline_snps",
+        type=int,
+        default=10,
+        help="number of snp supporting reads to count as a germline snp, for normal-in-tumor <-> normal matching",
+    )
+    parser.add_argument("--out_dir", type=str, required=True, help="output directory")
+    return parser
+
+
+def read_count(in_file: str) -> int:
+    with open(in_file, encoding="utf-8") as f:
+        return int(f.read().strip())
+
+
+class TrainingSetConsistency:
+    def __init__(  # pylint: disable=too-many-arguments
+        self,
+        target_crams: list[str],
+        normal_crams: list[str],
+        ground_truth_vcf: str,
+        hcr: str,
+        ref: str,
+        max_vars: int,
+        region: str,
+        min_count_snps: int,
+        min_count_germline_snps: int,
+        out_dir: str,
+        sp: SimplePipeline,
+    ):
+        self.target_crams = target_crams
+        self.normal_crams = normal_crams
+        self.ground_truth_vcf = ground_truth_vcf
+        self.hcr = hcr
+        self.ref = ref
+        self.max_vars = max_vars
+        self.region = region
+        self.out_dir = out_dir
+        self.min_count_snps = min_count_snps
+        self.min_count_germline_snps = min_count_germline_snps
+        self.sp = sp
+        os.makedirs(out_dir, exist_ok=True)
+
+    def check(self):
+        errors = []
+        infos = []
+        suspected_normal_in_tumor_crams = []
+        ground_truth_in_hcr = f"{self.out_dir}/ground_truth_in_hcr.bed"
+        ground_truth_to_check = f"{self.out_dir}/ground_truth_to_check.bed"
+        self.sp.print_and_run(
+            f"bedtools intersect -a {self.ground_truth_vcf} -b {self.hcr} -header > {ground_truth_in_hcr}"
+        )
+        self.vcf_to_bed(ground_truth_in_hcr, ground_truth_to_check, max_vars=self.max_vars)
+        hcr_in_region = f"{self.out_dir}/hcr_in_region.bed"
+        if self.region != "":
+            self.sp.print_and_run(f"echo {self.region} | sed 's/:/\t/' | sed 's/-/\t/' > {self.out_dir}/region.bed")
+            self.sp.print_and_run(
+                f"bedtools intersect -a {self.hcr} -b {self.out_dir}/region.bed | sort -k 1,1 -k 2,2n > {hcr_in_region}"
+            )
+        else:
+            self.sp.print_and_run(f"sort -k 1,1 -k 2,2n {self.hcr} > {hcr_in_region}")
+
+        # Validate that eacg target cram correlates or anti-correlates the ground-truth
+        for target_cram in self.target_crams:
+            hit_fraction = self.calc_hit_fraction(target_cram, ground_truth_to_check)
+            if hit_fraction < 0.95:
```

Comment:
A bit worried about this hardcoded number (recall in low AF CRAMs could be lower)

Reply:
Right. I already externalized it after it failed on FFPE